### DISCUSSION TITLE
Deployment, multiple domains, projects?

### DISCUSSION THREAD

**Discussion by MPatriot** [oqtane/oqtane.framework#1947] [Q&A] [answered]
I have multiple domain names pointing to the same server and until now I have been using 1 project to publish all the sites from.  This may be a problem however...
Currently its setup with:

localhost:44357 - default site, first alias
site1.com,www.site1.com - second site / alias
site2.com,www.site2.com - third site / alias

...all using the same tenant and project

It seems I almost need to have different Visual Studio projects for each domain so each can have its own localhost port

I first noticed this when trying to change the theme on site1.com and got the following error:

System.NullReferenceException: Object reference not set to an instance of an object.
   at Oqtane.Modules.Admin.Site.Index.SaveSite() in C:\Projects\IOFlowstane\oqtane.framework-master\Oqtane.Client\Modules\Admin\Site\Index.razor:line 476

This is looking for a file on my computer, not on the server, also, it is stated many times that certain actions can only be done in a development environment.  I wasn't aware of changing a theme as one of those things but if it is going to look on my computer than I suppose it would have to be.  Since multiple sites can't all use the same local host it seems like its necessary to have a different project for each domain, so each gets its own localhost port?

so then it could be:

localhost:portA,site1.com,www.site1.com 
localhost:portB,site2.com,www.site2.com
and if I wanted to I could have another project with its own localhost port just for playing around (default)

Does seem like how things have to be setup or can I do all of this from a single project?

Thanks,
Mark

**Comment by MPatriot** [comment]
The simplest way I found to have multiple sites in one project is to not have a localhost default project but to rename the default project to whatever the first first project is and then add its hosted aliases after local host.  Then, just change the localhost siteid to whatever site you are trying to edit in development environment.
Rather than go into the database to change this every time I want to work locally  on a different site I'll probably add it to my external module that follows the DNF example demonstrated on YouTube.
Also, I noticed that in order to edit a site locally it needs to be run locally with site settings or else clicking edit in the site management takes you to the hosted alias, not the localhost 

Cheers!

**Comment by leigh-pointer** [comment]
@MPatriot 

> System.NullReferenceException: Object reference not set to an instance of an object.
at Oqtane.Modules.Admin.Site.Index.SaveSite() in C:\Projects\IOFlowstane\oqtane.framework-master\Oqtane.Client\Modules\Admin\Site\Index.razor:line 476

> This is looking for a file on my computer, not on the server

This information is coming from the .pdb files which holds debugging and project state

**Reply by MPatriot** [reply]
Yeah, I don't know Leigh...
I seem to be having lots of problems publishing.  
As far as debugging goes, everything has been compiled and publish in release mode.
Everything seems to work fine on localhost but then when published the theme content and my DNF style module don't show up, though the code I made for a job in that that module does run on time.
I must not be deploying properly.  Was just publishing the Oqtane.Server project thinking everything else would go along with it

Thanks - Mark

**Comment by sbwalker** [comment, marked as answer]
@MPatriot it is clear what you are trying to do but in regards to development and aliases, I wanted to point out that Oqtane supports "virtual folders" as aliases which may be useful in your situation:

ie.

site1 = localhost:port#,www.site1.com
site2 = localhost:port#/site2,www.site2.com
site3 = localhost:port#/site3,www.site3.com

With this alias specification you can use the same localhost:port# for all 3 sites in Visual Studio development. 

Note that in version 3.0.2 it is now possible to specify which alias is the default for the site... and requests for non-default aliases will be redirected to the default. However the default alias will be different for your development environment than it is for your production environment so you will need to set them properly in each environment. You can access the site management UI directly in any environment by specifying the url of {alias}/admin/site ie. www.site1.com/admin/site

**Reply by MPatriot** [reply]
Thanks @sbwalker ,

Lot of good information there and more to learn and look into lol!

- Mark

**Reply by mixlplix** [reply]
So what is the point in having an alias if it always redirects to the default one?  We are now unable to work locally against our dev system in the cloud because it always redirects to the cloud.  Here is our setup.

**Reply by mixlplix** [reply]
<img width="653" alt="image" src="https://user-images.githubusercontent.com/1314264/173681203-b46e86f3-884f-4a4b-a0d3-e2c3775cdebf.png">

**Reply by sbwalker** [reply]
@mixlplix most of the configuration settings in Oqtane assume they are related to a specific environment. For example, SMTP settings, External Login, etc... So in the case of Aliases it is assumed that all aliases relate to the same environment. So if you have a production environment and you have configured domain.com and www.domain.com in your DNS then you would specify "domain.com,www.domain.com" as your aliases - and you would want to indicate that one of them is the default - otherwise search engines are going to heavily punish you for duplicate content. You could also include additional "vanity" domains which you use for external marketing purposes - but even in these cases you would still want to redirect to the default for the exact same reason. 

In your case it appears that you are trying to leverage the same development database for both localhost development as well as development environment testing - is this correct? From a purist perspective it would be more optimal to have a dedicated localhost environment and a dedicated development testing environment - each with their own database so that there is cleaner separation of concerns. That being said, there are always exceptions to the rule and I can understand that sometimes it is useful to be able to share databases between environments. The alias management area would need to be enhanced to support this specific scenario.

**Comment by sbwalker** [comment]
@MPatriot one other thing to keep in mind - which does not seem to be widely known - is that on a Windows machine you have the whole 127.x.x.x range at your disposal as a loopback address. So you can have 127.0.0.1 and 127.0.0.2, etc.. which can be used in Oqtane as unique Aliases to manage multiple sites.

**Comment by tvatavuk** [comment]
For local development (thx to @iJungleboy) one can also use `*.oqtane.me` because public DNS resolves it to local ip `127.0.0.1` eg.

site1.oqtane.me
site2.oqtane.me
site3.oqtane.me
...
![image](https://user-images.githubusercontent.com/2724850/153921130-2716d1b0-3277-4b45-ada8-a342ed4d7863.png)

For more info, please look https://oqtane.me/.

**Reply by ghhv** [reply]
Is there a trick to getting this working @tvatavuk ? 
I've tried editing launchsettings.json and starting VS as Administrator and checking firewall?
No matter what I do keep getting "Bad Request - Invalid Hostname  HTTP Error 400. The request hostname is invalid."

**Comment by tvatavuk** [comment]
With Oqtane Source Version, when is running from Visual Studio, by default it is using IISExpress and opens main oqtane site in browser as http://localhost:44357/

Than in Oqtane Sites Management we can create additional oqtane sites, and set additional websites bindings like:
- localhost:44357, site0.oqtane.me
- localhost:44357/site1, site1.oqtane.me
- localhost:44357/site2, site2.oqtane.me

![image](https://user-images.githubusercontent.com/2724850/158157299-54440eab-4b40-4457-94b9-9736eb22883d.png)

Last step is to config IISExpress to recognize additional bindings.

Please open `oqtane.framework-3.0.3\.vs\Oqtane\config\applicationhost.config` and find line 163

```xml
        <bindings>
          <binding protocol="http" bindingInformation="*:44357:localhost" />
        </bindings>
```
Insert bindings for `site0.oqtane.me`, `site1.oqtane.me` and `site2.oqtane.me`.

```xml
        <bindings>
          <binding protocol="http" bindingInformation="*:44357:localhost" />
          <binding bindingInformation="*:80:site0.oqtane.me" protocol="http" />
          <binding bindingInformation="*:80:site1.oqtane.me" protocol="http" />
          <binding bindingInformation="*:80:site2.oqtane.me" protocol="http" />
        </bindings>
```
Alternative is to use free 3rd party util to config IISExpress like https://www.jexusmanager.com/.

After this when you run Oqtane Source Code in IISEpress yuu can open additional URLs in your browser:

- http://site0.oqtane.me
- http://site1.oqtane.me
- http://site2.oqtane.me

You can find more info in video https://youtu.be/THg1gVAoh48 and recipe https://azing.org/oqtane/r/QOJI4-o3

Usefull links:
- https://oqtane.me/
- https://www.jexusmanager.com/
- https://github.com/jexuswebserver/JexusManager/releases

**Comment by mixlplix** [comment]
Got it.  That makes sense, but it’s also makes it a lot harder to develop when each developer has to create an identical copy of the nav and page setup of the dev/prod sites in order to mimic the behavior.

Then any new pages developed have to be manually created and the modules added to them when changes are deployed.

**Reply by sbwalker** [reply]
@mixlplix  Have you looked at the new Site Migration feature introduced in 3.1:

https://github.com/oqtane/Oqtane.Demo31

https://github.com/oqtane/Oqtane.Demo31/blob/master/Server/Manager/SiteMigration.1.0.1.cs

This allows you to create version-based content migrations for your site so that you can use CI/CD to deploy changes to various environments.

**Comment by mixlplix** [comment]
Thanks, that’s very helpful.

**Comment by sbwalker** [comment]
#2244 will be useful to those folks who need to define multiple primary aliases for a site in a particular environment:

![image](https://user-images.githubusercontent.com/4840590/174439543-8625f201-1313-46aa-86f0-f95f7a6a0277.png)

@mixlplix
